### ISSUE TITLE
Did you observe that the loss was stuck at the margin?

### ISSUE THREAD

**Issue by weedwind** [philipperemy/deep-speaker#9]
Hi, as mentioned in the title, I am also implementing deep speaker, using the same data as you used. I am wondering was your triplet loss able to get smaller than the margin value? It seems mine was stuck at the margin.

**Comment by philipperemy** [comment]
Yes this implementation is not yet functional. I think it's pretty hard to train a triplet loss.
I guess we have to go through supervised learning and train the network first.
Then, use those pre-trained weights and train the triplet loss.

**Comment by philipperemy** [comment]
@weedwind do you have some ideas on how to fix this?

**Comment by daniel-sc** [comment]
Hi everyone,
I'm not sure, if I'm correct, but it seems to me that the loss function is not really correct here. I tried to fix it - see https://github.com/philipperemy/deep-speaker/compare/master...daniel-sc:fixing_loss_function
But with this, the loss only declines only very little on training..

**Comment by philipperemy** [comment]
Thank you very much @daniel-sc !

I'm going to have a look at your code.

I'm not surprised that the loss only declines very little during training, if the dataset is truncated and if the samples are limited to only one second.

Have you tried training on the full dataset?

**Comment by daniel-sc** [comment]
Just realized that my new loss is wrong as well - must not collapse the vector to a scalar and replace square by sqrt in norm function - will update in the next days..

**Comment by philipperemy** [comment]
@daniel-sc ok got it! Looking forward to your new code!

**Comment by daniel-sc** [comment]
@philipperemy I created #12 with an additional commit, so the loss function should be correct there. But please double check it!

**Comment by linhdvu14** [comment]
@daniel-sc  Did your "deep speaker paper is wrong here, see e.g. face net paper" comment refer to FaceNet paper eq. (3) vs. Deep Speaker paper eq. (6)? 

If I'm reading correctly, sap/nap in Deep Speaker refers to cosine similarity while ||f(x^a) - f(x^p)|| / ||f(x^a) - f(x^n)|| refers to distance (i.e. dissimilarity). The equations are different interpretations of the same core idea, it seems.

Please correct me if I'm wrong.

**Comment by philipperemy** [comment]
@daniel-sc thanks I'll check it tomorrow!

**Comment by philipperemy** [comment]
Should be fixed with https://github.com/philipperemy/deep-speaker/pull/12
